src/lib.rs

```diff
@@ -222,35 +255,76 @@ where
             }
 
             State::ShortOptionCluster(_, rest) => {
-                let ch = rest.chars().next().unwrap();
-                let opt = Opt::Short(ch);
-                let rest = &rest[ch.len_utf8()..];
-                if rest.is_empty() {
-                    self.state = State::EndOfOption(opt);
-                } else {
+                let (opt, rest) = rest.consume_short_opt();
+                let opt = Opt::Short(opt);
+
+                if let Some(rest) = rest {
                     self.state = State::ShortOptionCluster(opt, rest);
+                } else {
+                    self.state = State::EndOfOption(opt);
                 }
+
                 Ok(Some(opt))
             }
 
-            State::LongOptionWithValue(opt, _) => Err(Error::DoesNotRequireValue(opt)),
+            State::LongOptionWithValue(opt, _) => {
+                self.state = State::Start;
+                Err(Error::DoesNotRequireValue(opt))
+            }
+
+            State::Positional(_) | State::PositionalOnly | State::End => Ok(None),
+        }
+    }
+
+    /// Stops the parsing of the current option. Namely, raises an error
+    /// if a value was provided that the option did not accept. This can
+    /// be used if you plan to call [`arg`][Options::arg] and want to
+    /// sacrifice a panic in [`Iterator::next`] for proper error
+    /// reporting if an option's parsing was not yet complete.
+    ///
+    /// This method never panics, and can also be used to continue
+    /// parsing options after `--` is seen. However, it is not a
+    /// substitute for [`restart`][Options::restart].
+    pub fn stop(&mut self) -> Result<A, ()> {
```

Comment:
> I don't really like the idea of needing to call `stop()` before parsing more options after the `--`.

The thing is, in order to support alternating arguments properly (i.e. `--arg value positional --other-arg -- positional --also-positional`), `Options` needs to remember that it has seen `--` and continue to return `None` even when it supposedly sees a new flag.

If you have any other ideas I'm all ears.

Reply:
> The thing is, in order to support alternating arguments properly (i.e. `--arg value positional --other-arg -- positional --also-positional`), `Options` needs to remember that it has seen `--` and continue to return `None` even when it supposedly sees a new flag.

That's a good point. To be honest, I never really thought about support alternating arguments. In this library's current state, alternating arguments are a bit difficult to deal with, since you'd basically need to check `next()` after each call to `arg()`.

Maybe we'd need to redesign the library's API to better support this...

Reply:
> That's a good point. To be honest, I never really thought about support alternating arguments. In this library's current state, alternating arguments are a bit difficult to deal with, since you'd basically need to check `next()` after each call to `arg()`.

I had actually added that to the documentation.

Reply:
I think we could just leave the API as it is for now -- only supporting options before positional arguments (there are some programs do this, like `cargo run` and `docker-compose exec`). So I would say keep `next()`'s behavior of parsing more options after `--` if it's called again after returning `None`.

The API would likely need to be redesigned to support options anywhere, and this `stop()` function is not a full solution.

Reply:
OK, I'll make those changes in a bit.

Reply:
Okay, so.

If `Options::next` returns `None` and `Options::arg` then returns `--`, you can detect the `--` since it gets returned, but that would be a pitfall for new users ("why does `--` show up as a positional argument?"). Instead, a method could be added that returns `true` if the last call to `Options::next` was because of `--`.

Reply:
I believe I've found a good compromise, marking this as resolved.